votor/src/certificate_pool_service.rs

```diff
@@ -0,0 +1,267 @@
+    pub(crate) own_vote_receiver: BLSVerifiedMessageReceiver,
+    pub(crate) bls_receiver: BLSVerifiedMessageReceiver,
+    pub(crate) event_sender: Sender<VotorEvent>,
+    pub(crate) commitment_sender: Sender<AlpenglowCommitmentAggregationData>,
+    pub(crate) certificate_sender: Sender<(CertificateId, CertificateMessage)>,
+}
+
+pub(crate) struct CertificatePoolService {
+    t_ingest: JoinHandle<()>,
+}
+
+impl CertificatePoolService {
+    pub(crate) fn new(ctx: CertificatePoolContext) -> Self {
+        let exit = ctx.exit.clone();
+        let t_ingest = Builder::new()
+            .name("solCertPoolIngest".to_string())
+            .spawn(move || {
+                if let Err(e) = Self::certificate_pool_ingest(ctx) {
+                    info!("Certificate pool service exited: {e:?}. Shutting down");
+                    exit.store(true, Ordering::Relaxed);
+                }
+            })
+            .unwrap();
+
+        Self { t_ingest }
+    }
+
+    // TODO: properly bubble up errors
+    fn certificate_pool_ingest(mut ctx: CertificatePoolContext) -> Result<(), RecvError> {
+        let mut events = vec![];
+        let mut cert_pool = certificate_pool::load_from_blockstore(
+            &ctx.my_pubkey,
+            &ctx.root_bank_cache.root_bank(),
+            ctx.blockstore.as_ref(),
+            Some(ctx.certificate_sender.clone()),
+            &mut events,
+        );
+
+        // Wait until migration has completed
+        info!("{}: Certificate pool loop initialized", &ctx.my_pubkey);
+        Votor::wait_for_migration_or_exit(&ctx.exit, &ctx.start);
+        info!("{}: Certificate pool loop starting", &ctx.my_pubkey);
+        let mut current_root = ctx.root_bank_cache.root_bank().slot();
+
+        // Standstill tracking
+        let mut standstill_timer = Instant::now();
+        let mut highest_finalized_slot = cert_pool.highest_finalized_slot();
+
+        // Kick off parent ready
+        let root_bank = ctx.root_bank_cache.root_bank();
+        let root_block = (
+            root_bank.slot(),
+            root_bank.block_id().unwrap_or_default(),
+            root_bank.hash(),
+        );
+        let mut highest_parent_ready = root_bank.slot();
+        events.push(VotorEvent::ParentReady {
+            slot: root_bank.slot().checked_add(1).unwrap(),
+            parent_block: root_block,
+        });
+
+        // Ingest votes into certificate pool and notify voting loop of new events
+        while !ctx.exit.load(Ordering::Relaxed) {
+            // TODO: we need set identity here as well
+            Self::add_produce_block_event(
+                &mut highest_parent_ready,
+                &cert_pool,
+                &mut ctx,
+                &mut events,
+            );
+
+            if standstill_timer.elapsed() > STANDSTILL_TIMEOUT {
+                events.push(VotorEvent::Standstill(highest_finalized_slot));
```

Comment:
Can we just use cert_pool.highest_finalized_slot() here?
I think this highest_finalized_slot is just used to:
1. check whether the returned `highest_finalized_slot` is larger than the local copy, which `certificate_pool` already checks before returning
2. used in Standstill event here, which we can get from `cert_pool.highest_finalized_slot()`.

Reply:
great idea, I think that should work.